CRM/ManualDirectDebit/Common/MandateStorageManager.php

```diff
@@ -201,87 +201,97 @@ private function setMandateForCreatingDependency($mandateId) {
-     * so we assign them to currently created mandate,
-     * and all next contributions we will assign in manualdirectdebit_civicrm_postSave_civicrm_membership_payment hook
-     *
-     *
+     * If DirectDebit was installed first, it gets last created contribution id for current user
      */
-    if ($this->isDirectDebitInstalledFirst()) {
+    if ($this->isDirectDebitInstalledBeforeMembershipExtras()) {
       $sqlContributionID = "SELECT MAX(`id`) AS id FROM civicrm_contribution WHERE `contact_id` = %1";
       $queryResult = CRM_Core_DAO::executeQuery($sqlContributionID, [
         1 => [
-          2,
+          $contactId,
           'String',
         ],
       ]);
       $queryResult->fetch();
-      $lastContributionID = $queryResult->id;
-
-      $this->assignRecurringContributionMandate($lastInsertedRecurrContribution, $mandateId);
-      $this->assignContributionMandate($lastContributionID, $mandateId);
-    }
-    else {
+      $contributionIds[] = $queryResult->id;
+    } else {
 
       /**
-       *
-       *
-       * If Membership was installed first, at this moment it already divide contributions in correct amount of instalments.
-       * And assign all this contributions to 'civicrm_membership_payment'
-       * So at this time we have all necessary data, and manualdirectdebit_civicrm_postSave_civicrm_membership_payment hook
-       * will no longer execute
-       * so we assign last inserted recurr contribution to currently created mandate,
-       * and to all contributions which was assign to this recurr contribution
-       *
-       *
+       * If Membership was installed first, it gets all contributions for last inserted recurring contribution
+       * for current contact
        */
       $allContributionIdsForRecurr = civicrm_api3('Contribution', 'get', [
         'sequential' => 1,
         'return' => ["id"],
         'contribution_recur_id' => $lastInsertedRecurrContribution,
       ]);
 
-      $contributionIds = [];
       foreach ($allContributionIdsForRecurr['values'] as $value) {
         $contributionIds[] = $value['contribution_id'];
       }
 
-      foreach ($contributionIds as $contributionId) {
-        $this->assignContributionMandate($contributionId, $mandateId);
-      }
-
-      $this->assignRecurringContributionMandate($lastInsertedRecurrContribution, $mandateId);
     }
+    return $contributionIds;
   }
 
-  private function isDirectDebitInstalledFirst() {
-    $sqlManualdirectdebitId3 = "SELECT id FROM civicrm_extension WHERE `full_name` = 'uk.co.compucorp.manualdirectdebit'";
-    $manualdirectdebitIdQueryResult = CRM_Core_DAO::executeQuery($sqlManualdirectdebitId3);
+  /**
+   * Checks if 'Manual Direct Debit' extension was installed before 'Membership
+   * Exras'
+   */
+  private function isDirectDebitInstalledBeforeMembershipExtras() {
+    $sqlManualDirectDebitId = "SELECT id FROM civicrm_extension WHERE `full_name` = 'uk.co.compucorp.manualdirectdebit'";
```

Comment:
I looked a little deeper into this, and found the list of installed modules is queried from the database without an `ORDER BY` clause. This is very problematic, because basically, there is no guarantee on the order on which extension hooks are executed. Normally it should return records in the order they were inserted, but again, there is no guarantee this will always happen, and since we're dealing with money matters here, I think we should strive to guarantee a consistent behaviour.

https://github.com/civicrm/civicrm-core/blob/3d04355de56e2293afda299bc4e00d5816010082/CRM/Extension/Mapper.php#L291-L296

Maybe we could try to push a PR for CiviCRM core to add an `ORDER BY` clause, to eliminate this risk. I think CiviCRM core team is striving to move to a single event managing system where events may be weighted, but in the meantime, I think this would be a good improvement. We can also implement some other way to tell if membershipextras hook has already run. Or maybe add the recurring contribution ID and contactID to a queue that is processed asynchronously.

Maybe the simplest way would be to check:
1. If membershipextras extension is installed
2. If the number of existing contributions equals the number of installments for the recurring contribution

And assign the mandate to the contributions accordingly.

Eg:
```php
$recurringContribution = $this->getRecurringContribution($recurringContributionID);

if ($this->isMembershipExtrasInstalled() && $recurringContribution->installments == $this->countRecurringContributionPayments($recurringContributionID)) {
  // Process as if membership extras has already been run, ie. return list of contributions associated to recurring contribution.
  ...
} else {
  // Process as if DD extension ran before membershipextras, ie. return last contribution created for the contact
  ..
}
```

CC @guanhuan